official/utils/testing/golden.py

```diff
@@ -0,0 +1,335 @@
+      tf_version_reference, tf_git_version_reference = json.load(f)  # pylint: disable=unpacking-non-sequence
+
+    tf_version_comparison = ""
+    if tf.GIT_VERSION != tf_git_version_reference:
+      tf_version_comparison = (
+          "Test was built using:     {} (git = {})\n"
+          "Local TensorFlow version: {} (git = {})"
+          .format(tf_version_reference, tf_git_version_reference,
+                  tf.VERSION, tf.GIT_VERSION)
+      )
+
+    with self.test_session(graph=graph) as sess:
+      sess.run(init)
+      try:
+        saver.restore(sess=sess, save_path=os.path.join(
+            data_dir, self.ckpt_prefix))
+        if differences:
+          print()
+          tf.logging.warn(
+              "The provided graph is different than expected:\n  {}\n"
+              "However the weights were still able to be loaded.\n{}".format(
+                  differences, tf_version_comparison)
+          )
+      except:  # pylint: disable=bare-except
+        raise self.failureException(
+            "Weight load failed. Graph comparison:\n  {}{}"
+            .format(differences, tf_version_comparison))
+
+      eval_results = [op.eval() for op in ops_to_eval]
+      if correctness_function is not None:
+        results = correctness_function(*eval_results)
+        with open(os.path.join(data_dir, "results.json"), "rt") as f:
+          expected_results = json.load(f)
+        self.assertAllClose(results, expected_results)
+
+  def _save_or_test_ops(self, name, graph, ops_to_eval=None, test=True,
+                        correctness_function=None):
+    """Utility function to automate repeated work of graph checking and saving.
+
+    The philosophy of this function is that the user need only define ops on
+    a graph and specify which results should be validated. The actual work of
+    managing snapshots and calculating results should be automated away.
+
+    Args:
+      name: String defining the run. This will be used to define folder names
+        and will be used for random seed construction.
+      graph: The graph in which the test is conducted.
+      ops_to_eval: Ops which the user wishes to be evaluated under a controlled
+        session.
+      test: Boolean. If True this function will test graph correctness, load
+        weights, and compute numerical values. If False the necessary test data
+        will be generated and saved.
+      correctness_function: This function accepts the evaluated results of
+        ops_to_eval, and returns a list of values. This list must be JSON
+        serializable; in particular it is up to the user to convert numpy
+        dtypes into builtin dtypes.
+    """
+
+    ops_to_eval = ops_to_eval or []
+
+    if test:
+      try:
+        self._evaluate_test_case(
+            name=name, graph=graph, ops_to_eval=ops_to_eval,
+            correctness_function=correctness_function
+        )
+      except:
+        tf.logging.error("Failed unittest {}".format(name))
```

Comment:
In google3, we usually use %s as place holder and rely on default format to convert the message, eg ("Failed unittest %s", name). I guess your code is also fine, but just need to be sure for consistency.

Reply:
Heh. I always want to nit people to replace all the `%s` with `{}`, but I fight the urge. I prefer the bracket style personally, but accept whatever everyone decides here.